core/embed/rust/src/ui/util.rs

```diff
@@ -31,6 +34,20 @@ pub fn u32_to_str(num: u32, buffer: &mut [u8]) -> Option<&str> {
     }
 }
 
+pub fn zeroize_vec<T, const N: usize>(val: &mut Vec<T, N>)
+where
+    T: DefaultIsZeroes,
+{
+    let slice: &mut [T] = val.as_mut();
```

Comment:
This only gets the slice spanning the part of the buffer that is currently filled, i.e. typing your pin and backspacing it will result in it staying in the buffer after zeroize.

You can get the whole buffer with
```rust
let ptr = val.as_mut_ptr();
let slice = core::slice::from_raw_parts_mut(ptr, N * core::mem::size_of::<T>());
```
however `from_raw_parts_mut` requires that

> `data` must point to `len` consecutive properly initialized values of type `T`

and the values in heapless::Vec start as uninitialized. Is this to ensure that destructors are called upon overwriting the data (and therefore safe for u8 which doesn't have destructors), or could this still result in UB?

I realize this is rather obscure corner case, just curious what would be the right way to solve it.

Reply:
one thing comes to mind is to support this only for `Vec<T: Default>` and manually fill the vec up to capacity before zeroizing

of course, a bad thing about all of this is the generic explosion

Reply:
TBH we'll be fine with just `Vec<u8>`

Reply:
done in [1878b46](https://github.com/trezor/trezor-firmware/pull/2552/commits/1878b46b89510a06d16de8ec02b7db98c5b9784d)